### PR TITLE
Cmake yaml module

### PR DESCRIPTION
**Pull request by tejlmand**
This PR introduces a CMake yaml module to facilitate reading and writing of simple yaml files directly in CMake.

Zephyr uses yaml intensively, and in CMake there is already several places where basic yaml files are written.

Having a common module helps us to cleanup and align yaml usage in Zephyr CMake and make future use-cases easier to implement.

### CODE REVIEW

File: cmake/modules/yaml.cmake [hunk 1]
```diff
@@ -0,0 +1,403 @@
...
+  cmake_parse_arguments(ARG_YAML "" "FILE;NAME" "" ${ARGN})
+
+  zephyr_check_arguments_required_all(${CMAKE_CURRENT_FUNCTION} ARG_YAML FILE NAME)
+  internal_yaml_context_free(NAME ${ARG_YAML_NAME})
+
+  zephyr_create_scope(${ARG_YAML_NAME})
+  zephyr_set(FILE ${ARG_YAML_FILE} SCOPE ${ARG_YAML_NAME})
+
+  execute_process(COMMAND python -c
+    "import json; import yaml; print(json.dumps(yaml.safe_load(open('${ARG_YAML_FILE}'))))"
```
Comment (nordicjm) [inline cmake/modules/yaml.cmake:149]: (not requesting change) something to probably thing about is that this escapes from of cmake into python as another process, from a security perspective anyone can put any file that is visible to the OS and might allow some level of privileged file access

File: cmake/modules/yaml.cmake [hunk 2]
```diff
@@ -0,0 +1,403 @@
...
+  string(JSON json_content REMOVE "${json_content}" ${ARG_YAML_KEY})
+
+  zephyr_set(JSON "${json_content}" SCOPE ${ARG_YAML_NAME})
+endfunction()
+
+# Usage:
+#   yaml_write(NAME <name>)
+#
+# Write the YAML context <name> to the file which were given with the earlier
+# yaml_load() or yaml_create() call.
```
Comment (nordicjm) [inline cmake/modules/yaml.cmake:342]:
```suggestion
# 'yaml_load()' or 'yaml_create()' call.
```

File: cmake/modules/yaml.cmake (hunk 1 above)
Comment (tejlmand) [reply cmake/modules/yaml.cmake:149]: The `ARG_YAML_FILE` is passed as an argument to `yaml_load()`, so true that a caller of said function can do:
`yaml_load(NAME foo FILE "<insert python escape command>")`.

But if the caller of `yaml_load()` can do that, then that caller can much easier call `execute_process()` directly without the hassle of going through another function. Just call: `execute_process(COMMAND python -c "...")` directly.

So I guess you're basically saying that CMake is unsafe, especially if you use `sudo cmake` :wink:

File: cmake/modules/yaml.cmake [hunk 3]
```diff
@@ -0,0 +1,403 @@
...
+
+# Usage:
+#   yaml_write(NAME <name>)
+#
+# Write the YAML context <name> to the file which were given with the earlier
+# 'yaml_load()' or 'yaml_create()' call.
+#
+# NAME <name>: Name of the YAML context
+#
+function(yaml_write)
```
Comment (pdgendt) [inline cmake/modules/yaml.cmake:346]: Wouldn't it make sense to decouple the context from the file? And add a `yaml_save` function that can be passed a filename?

Comment (tejlmand) [reply cmake/modules/yaml.cmake:346]: I guess it depends on the use-case.
The two use-cases I currently had in mind were
1. you know everything at one location, like the [domains.yaml](https://github.com/zephyrproject-rtos/zephyr/blob/main/share/sysbuild/cmake/domains.cmake) file.
    There it doesn't really matter.
2. You may know about the file (and load it into a context) at one location, but you'll be updating the context at other locations, then all you want to know is the agreed context name so that you can. You don't want/need to know about file names. That's why the actual filename is a property on the context.

but I see your proposal / use-case as just as valid, so perhaps allowing for either write to specific filename, or omitting the filename when creating a new context.

In short:
- allow specifying a file name on `yaml_write()`
  The filename property of the context will be used when filename is not provided with `yaml_write`.
    - Fail with error if calling `yaml_write()` if context has no filename property defined and filename was not passed as argument.
- make filename optional on `yaml_create()`

Comment (pdgendt) [reply cmake/modules/yaml.cmake:346]: I think these additions are welcome, can we rename `yaml_write` to `yaml_save`? If you use `load`, then `save` seems better suited.

Would it also make sense to allow loading multiple files to amend/overwrite values in the current context? For this the decoupling makes over more sense.

Comment (tejlmand) [reply cmake/modules/yaml.cmake:346]: > can we rename yaml_write to yaml_save? If you use load, then save seems better suited.

Good point and I agree, will do.

> Would it also make sense to allow loading multiple files to amend/overwrite values in the current context? For this the decoupling makes over more sense.

I would leave that for a future enhancement.
But one use-case (not yet supported) which I consider useful, is to be able to create a new context and then merge that into a context loaded from file. That principle is very similar to the loading of multiple files.

Comment (tejlmand) [reply cmake/modules/yaml.cmake]: > I think these additions are welcome, can we rename yaml_write to yaml_save?

Done

File: cmake/modules/yaml.cmake [hunk 4]
```diff
@@ -0,0 +1,414 @@
...
+  if(TARGET ${ARG_YAML_NAME}_scope)
+    list(POP_FRONT ARG_YAML_UNPARSED_ARGUMENTS out-var)
+    set(${out-var} TRUE PARENT_SCOPE)
+  else()
+    set(${out-var} ${ARG_YAML_NAME}-NOTFOUND PARENT_SCOPE)
+  endif()
+endfunction()
+
+# Usage:
+#   yaml_create(FILE <file> NAME <name>)
```
Comment (pdgendt) [inline cmake/modules/yaml.cmake:105]: `File <file>` is optional now.

File: cmake/modules/yaml.cmake [hunk 5]
```diff
@@ -0,0 +1,414 @@
...
+endfunction()
+
+# Usage:
+#   yaml_save(NAME <name>)
+#
+# Write the YAML context <name> to the file which were given with the earlier
+# 'yaml_load()' or 'yaml_create()' call.
+#
+# NAME <name>: Name of the YAML context
+#
```
Comment (pdgendt) [inline cmake/modules/yaml.cmake:347]: Missing `File` argument info.

File: cmake/modules/yaml.cmake (hunk 4 above)
Comment (tejlmand) [reply cmake/modules/yaml.cmake]: Fixed

File: cmake/modules/yaml.cmake (hunk 5 above)
Comment (tejlmand) [reply cmake/modules/yaml.cmake]: Fixed

**Review by pdgendt — DISMISSED** [review]

**Review by pdgendt — DISMISSED** [review]

**Review by pdgendt — DISMISSED** [review]

**Review by pdgendt — APPROVED** [review]

**Review by nordicjm — APPROVED** [review]
